lib-bucket-provisioner CRD-only v1.0.0 (openshift)

guymguym opened the pull request:
See #656.

Thanks submitting your Operator. Please check below list before you create your Pull Request.

### New Submissions

* [x] Have you selected the Project *Community Operator Submissions* in your PR on the right-hand menu bar?
* [x] Are you familiar with our [contribution guidelines](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md)?
* [x] Have you [packaged and deployed](https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md) your Operator for Operator Framework?
* [x] Have you tested your Operator with all Custom Resource Definitions?
* [x] Have you tested your Operator in all supported [installation modes](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Documentation/design/building-your-csv.md#operator-metadata)?

### Updates to existing Operators

* [ ] Is your new CSV pointing to the previous version with the `replaces` property?
* [x] Is your new CSV referenced in the [appropriate channel](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#bundle-format) defined in the `package.yaml` ?
* [ ] Have you tested an update to your Operator when deployed via OLM?

### Your submission should not

* [x] Modify more than one operator
* [x] Submit operators to both `upstream-community-operators` and `community-operators` at once
* [x] Modify any files outside the above mentioned folders
* [x] Contain more than one commit. **Please squash your commits.**

### Operator Description must contain (in order)

1. [x] Description about the managed Application and where to find more information
2. [x] Features and capabilities of your Operator and how to use it
3. [x] Any manual steps about potential pre-requisites for using your Operator

### Operator Metadata should contain

* [x] Human readable name and 1-liner description about your Operator
* [x] Valid [category name](https://github.com/operator-framework/community-operators/blob/master/docs/required-fields.md#categories)<sup>1</sup>
* [x] One of the pre-defined [capability levels](https://github.com/operator-framework/operator-courier/blob/4d1a25d2c8d52f7de6297ec18d8afd6521236aa2/operatorcourier/validate.py#L556)<sup>2</sup>
* [x] Links to the maintainer, source code and documentation
* [x] Example templates for all Custom Resource Definitions intended to be used
* [x] A quadratic logo

Remember that you can preview your CSV [here](https://operatorhub.io/preview).

--

<sup>1</sup> If you feel your Operator does not fit any of the pre-defined categories, file a PR against this repo and explain your need

<sup>2</sup> For more information see [here](https://github.com/operator-framework/operator-sdk/blob/master/doc/images/operator-capability-level.svg)

guymguym commented:
Added CRD validations, hope now scorecard will pass...

guymguym commented:
Not sure why it failed now, I added CRD validations but scorecard still says:
```
Basic Tests:
	Spec Block Exists: 1/1
	Status Block Exists: 0/1
	Writing into CRs has an effect: 0/1
OLM Tests:
	Provided APIs have validation: 0/0
	Owned CRDs have resources listed: 1/1
	CRs have at least 1 example: 1/1
	Spec fields with descriptors: 0/5
	Status fields with descriptors: 0/0

Total Score: 40%
SUGGESTION:Add a 'status' field to your Custom Resource
SUGGESTION:The operator should write into objects to update state. No PUT or POST requests from the operator were recorded by the scorecard.
SUGGESTION:Add a spec descriptor for versioned
SUGGESTION:Add a spec descriptor for SSL
SUGGESTION:Add a spec descriptor for additionalConfig
SUGGESTION:Add a spec descriptor for generateBucketName
SUGGESTION:Add a spec descriptor for storageClassName
```

Any help would be much appreciated

guymguym commented:
@J0zi Hey, any idea what am I missing here?

guymguym commented:
@dmesser I need some help here, who can assist? Thanks

J0zi commented:
@guymguym you listed all issues from scorecard. You don`t need to meet all criteria. Firstly, you can add a 'status' field to your Custom Resource. This test checks the Custom Resource(s) created in the cluster to make sure that all CRs have a status block. We will see if it is green then.

If not, second issue is - The operator should write into objects to update state. No PUT or POST requests from the operator were recorded by the scorecard. But firstly, just add status field and we will see.

guymguym commented:
Thanks @J0zi - You mean that I add a `status: {}` to the CSV `alm-examples`?

guymguym commented:
@J0zi Weird - it still says that I don't have status after I added.
Also why does it say that I don't have spec descriptors??
I already added it..
It seems like it doesn't see my changes? or maybe something is off with my PR?


```
Basic Tests:
	Spec Block Exists: 1/1
	Status Block Exists: 0/1
	Writing into CRs has an effect: 0/1
OLM Tests:
	Provided APIs have validation: 0/0
	Owned CRDs have resources listed: 1/1
	CRs have at least 1 example: 1/1
	Spec fields with descriptors: 0/5
	Status fields with descriptors: 0/0

Total Score: 40%
SUGGESTION:Add a 'status' field to your Custom Resource
SUGGESTION:The operator should write into objects to update state. No PUT or POST requests from the operator were recorded by the scorecard.
SUGGESTION:Add a spec descriptor for additionalConfig
SUGGESTION:Add a spec descriptor for generateBucketName
SUGGESTION:Add a spec descriptor for storageClassName
SUGGESTION:Add a spec descriptor for versioned
SUGGESTION:Add a spec descriptor for SSL
```

J0zi commented:
@guymguym you can compare with operators where Status block tests passed. See https://github.com/operator-framework/community-operators/pull/619 and https://github.com/operator-framework/community-operators/pull/666

J0zi commented:
> not running the CI
@guymguym you are looking on debug output from Travis, but Travis is not running for OpenShift. As you see, your operator has issues with CVP/operator-scorecard-tests and error is that Status block is missing.

What I'm trying to say, is that #619 has no problem with Status block, so you can check how is Status block defined in ClusterServiceVersion and compare with yours. It could help you pass scorecard tests.

J0zi commented:
/test all

J0zi commented:
/retest CVP

J0zi commented:
/test all

J0zi commented:
hanged CICD pod

dmesser commented:
/test

J0zi commented:
@guymguym, I decided to merge it as I see your status field in every crd. However, consider adding 'replaces' field. Or it is just text change?

J0zi commented:
@guymguym: Important announcement: Flat operator directory structure is obsolete from 23-rd of October 2019, only nested directory structure will be accepted. You can use nested structure before this date also.

jozzi commented:
/test

jozzi commented:
/retest CVP

jozzi commented:
/ok-to-test

jozzi commented:
/test all

guymguym commented:
Hi @J0zi @dmesser 
I updated this PR and it still fails in CI - probably because this is a CRD-only CSV.
Can you let me know if something is missing?
Thanks

J0zi commented:
@guymguym, consider adding specDescriptors and statusDescriptors to your CSV.

guymguym commented:
@J0zi Thanks I will certainly consider it for next release.
Is it mandatory, or can we push it as is to unblock https://github.com/openshift/ocs-operator/pull/360 and I will add later?

J0zi commented:
@guymguym, the last question, is all working and tested?

guymguym commented:
Yes - I tested using the following:
```
$ minikube start
$ minikube mount /Users:/Users &
$ make operator.test OP_PATH=community-operators/lib-bucket-provisioner VERBOSE=1
Pulling docker image                              [  Processing  ]
Using default tag: latest
latest: Pulling from operator-framework/operator-testing
9d48c3bd43c5: Pull complete
7f94eaf8af20: Pull complete
9fe9984849c1: Pull complete
e17f0ad88222: Pull complete
e8bd20cf6e8a: Pull complete
dbbc603b9086: Pull complete
17443955f0b9: Pull complete
2131a647cda4: Pull complete
237e1e9be13e: Pull complete
19d8b9bd6175: Pull complete
2c9037065787: Pull complete
1086a2761f1c: Pull complete
bdde2603580a: Pull complete
3bd708cd73c2: Pull complete
fef40e671ab8: Pull complete
Digest: sha256:d65875c666b0a85e1319d883c87d489a512ddc238167d627931d0fd570b49863
Status: Downloaded newer image for quay.io/operator-framework/operator-testing:latest
quay.io/operator-framework/operator-testing:latest
Pulling docker image                              [  OK  ]
Find kube config                                  [  /Users/gu/.kube/config  ]
Find kube context                                 [  minikube   ]
Try kube master                                   [  Trying 192.168.64.91:8443  ]
Try kube master                                   [  192.168.64.91:8443  ]
Operator version detected                         [  1.0.0  ]
Creating namespace                                [  Processing  ]
Creating namespace                                [  OK  ]
Verify operator                                   [  Processing  ]
Verify operator                                   [  OK  ]
Install OLM                                       [  Processing  ]
Install OLM                                       [  OK  ]
Make registry configmap                           [  Processing  ]
    creating subscription files                   [  OK  ]
    creating operator group file                  [  OK  ]
    creating CR file                              [  OK  ]
    creating kubeconfig secret file               [  OK  ]
    creating kubeconfig volume file               [  OK  ]
    creating kubeconfig secret mount              [  OK  ]
    creating config map registry                  [  OK  ]
lib-bucket-provisioner.catalogsource.yaml
Make registry configmap                           [  OK  ]
Operator deployment                               [  Processing  ]
    Apply OPERATOR GROUP file                     [  OK  ]
    Applying object to cluster                    [  Processing  ]
secret/scorecard-kubeconfig created
operatorgroup.operators.coreos.com/lib-bucket-provisioner-og created
configmap/lib-bucket-provisioner created
catalogsource.operators.coreos.com/lib-bucket-provisioner-ocs created
subscription.operators.coreos.com/lib-bucket-provisioner-sub created
    Applying object to cluster                    [  OK  ]
    Checking subscriptions if passes              [  Processing  ]
Checking subscription "lib-bucket-provisioner-sub" reaches "AtLatestKnown" state
retrying check of subscription status for "lib-bucket-provisioner-sub"...
subscription status for "lib-bucket-provisioner-sub": null
subscription status for "lib-bucket-provisioner-sub": "UpgradePending"
subscription status for "lib-bucket-provisioner-sub": "AtLatestKnown"
Successfully installed subscription "lib-bucket-provisioner-sub" in namespace "lib-bucket-provisioner-yujlt"
    Checking subscriptions if passes              [  OK  ]
    Checking csv if passes                        [  Processing  ]
Check that clusterserviceversion "lib-bucket-provisioner.v1.0.0" status is set to "Succeeded"
clusterserviceversion phase for "lib-bucket-provisioner.v1.0.0": "Succeeded"
Check that clusterserviceversion "lib-bucket-provisioner.v1.0.0" in second namespace status is set to "Succeeded"
Successfully installed clusterserviceversion "lib-bucket-provisioner.v1.0.0" in namespace "lib-bucket-provisioner-yujlt"
    Checking csv if passes                        [  OK  ]
    Waiting for deployment                        [  Processing  ]
deployment "lib-bucket-provisioner" successfully rolled out
Successfully rolled out deployment "lib-bucket-provisioner" in namespace "lib-bucket-provisioner-yujlt"
    Waiting for deployment                        [  OK  ]
Operator deployment                               [  OK  ]
Test operator with scorecard                      [  Processing  ]
Running scorecard trough all CR


Running operator-sdk scorecard against /tmp/tmp.dBnDIB/deploy/lib-bucket-provisioner/1.0.0/lib-bucket-provisioner.v1.0.0.clusterserviceversion.yaml with /tmp/tmp.dBnDIB/deploy/crs/XXXXfHjfeb.cr.yaml
DEBU[0000] Debug logging is set
WARN[0000] Could not load config file; using flags
WARN[0000] Plugin directory not found; skipping external plugins: stat scorecard: no such file or directory
Basic Tests:
	Spec Block Exists: 1/1
	Status Block Exists: 0/1
	Writing into CRs has an effect: 0/1
OLM Tests:
	Provided APIs have validation: 0/0
	Owned CRDs have resources listed: 1/1
	CRs have at least 1 example: 1/1
	Spec fields with descriptors: 0/5
	Status fields with descriptors: 0/0

Total Score: 40%
SUGGESTION: Add a 'status' field to your Custom Resource
SUGGESTION: The operator should write into objects to update state. No PUT or POST requests from the operator were recorded by the scorecard.
SUGGESTION: Add a spec descriptor for reclaimPolicy
SUGGESTION: Add a spec descriptor for storageClassName
SUGGESTION: Add a spec descriptor for additionalState
SUGGESTION: Add a spec descriptor for claimRef
SUGGESTION: Add a spec descriptor for endpoint
PASS: Scorecard test passed
NORMAL
Printing operator logs
Logs for container lib-bucket-provisioner in pod lib-bucket-provisioner:

configmap "lib-bucket-provisioner" deleted
secret "scorecard-kubeconfig" deleted
subscription.operators.coreos.com "lib-bucket-provisioner-sub" deleted
operatorgroup.operators.coreos.com "lib-bucket-provisioner-og" deleted
clusterserviceversion.operators.coreos.com "lib-bucket-provisioner.v1.0.0" deleted
customresourcedefinition.apiextensions.k8s.io "objectbucketclaims.objectbucket.io" deleted
customresourcedefinition.apiextensions.k8s.io "objectbuckets.objectbucket.io" deleted
Test operator with scorecard                      [  OK  ]
```

guymguym commented:
NOTE: I just updated all installModes to supported: true after a discussion with @awgreene
